src/lib/tecs/TECS.cpp

```diff
@@ -452,6 +452,38 @@ void TECS::_updateTrajectoryGenerationConstraints()
 	_velocity_control_traj_generator.setMaxVelDown(_max_climb_rate); // different convention for FW than for MC
 }
 
+void TECS::_updateFlightPhase(float altitude_sp_amsl, float height_rate_setpoint)
+{
+	const bool input_is_height_rate = PX4_ISFINITE(height_rate_setpoint);
+
+	// update flight phase state (are we flying level, climbing or descending)
+	if (input_is_height_rate) {
+
+		if (PX4_ISFINITE(_velocity_control_traj_generator.getCurrentPosition())) {
+			// we have a valid altitude setpoint which means that we are flying level
+			_flight_phase = tecs_status_s::TECS_FLIGHT_PHASE_LEVEL;
+
+		} else if (_velocity_control_traj_generator.getCurrentVelocity() > FLT_EPSILON) {
+			_flight_phase = tecs_status_s::TECS_FLIGHT_PHASE_CLIMB;
+
+		} else {
+			_flight_phase = tecs_status_s::TECS_FLIGHT_PHASE_DESCEND;
+		}
+
+	} else {
+		// stay in flight phase level if only small altitude changes are demanded (<10m)
+		if (altitude_sp_amsl - _alt_control_traj_generator.getCurrentPosition() > 10.0f) {
```

Comment:
Instead of hard-coding the limits (10m atm), we could link it to FW_ALT_ERR_U and FW_ALT_ERR_O